### PR TITLE
Make build process faster

### PR DESCRIPTION
**Pull request by mrks**
TODOs:
- [x] add link to blog post in README.md
- [x] check if ccache still makes sense, [find fix for pch issue](https://stackoverflow.com/questions/59095795/gcc-precompiled-header-not-deterministic)
- [x] adapt new_ naming according to whatever we decide in #1877 
- [x] solve JIT issues
- [x] get cmake 3.16 final into the CI image
- [x] verify that asan/ubsan/tsan still work
- [x] decide on which CI steps should use unity builds
- [x] check performance 
- [x] Once unmodified branch runs on Catalina, fix Mac CI build
- [x] Update Docker image with gcc 9, clang, cmake
- [x] Add notice for non-3.16 / non unity build users to CMakeLists.txt
- [x] Have a look at boost warning in cmake
- [x] check if hyrise_impl.a as added by #1474 changes the compile time
- [x] Update cmake on CI Mac (https://github.com/Homebrew/homebrew-core/pull/47226)
- [x] increase ccache folder size
- [x] Add documentation for developers (sloppiness, unity)
- [x] self-review
- [x] Feedback to cmake on unity+clang-tidy
- [x] better describe ccache options

* Keeps the build process from waiting on jemalloc
* Uses -O0 for debug linking
* Enables the use of unity builds (requires cmake 3.16)
* Enables precompiled-headers (requires cmake 3.16)
* Enables dynamic linking for Hyrise
* Uses LLVM lld instead of ld under Linux

Additionally
* Fixes some things that will become warnings with clang 9
* Fixes some ODR violations

compiletime

For details see this blog post: https://medium.com/p/8523135aed72

Performance:
```diff
 +----------------+----------------+-------+---------------+-------+------------+---------+
 | Benchmark      | prev. iter/s   | runs  | new iter/s    | runs  | change [%] | p-value |
 +----------------+----------------+-------+---------------+-------+------------+---------+
 | TPC-H 01       | 0.599761068821 | 36    | 0.60238713026 | 37    | +0%        |  0.5546 |
 | TPC-H 02       | 11.8571109772  | 712   | 12.149933815  | 729   | +2%        |  0.0000 |
 | TPC-H 03       | 6.96436452866  | 418   | 6.91861820221 | 416   | -1%        |  0.1704 |
 | TPC-H 04       | 6.85212564468  | 412   | 6.93864488602 | 417   | +1%        |  0.0036 |
 | TPC-H 05       | 4.43019294739  | 266   | 4.34368038177 | 261   | -2%        |  0.0013 |
 | TPC-H 06       | 195.011672974  | 11701 | 187.01574707  | 11221 | -4%        |  0.0000 |
 | TPC-H 07       | 7.42980623245  | 446   | 7.67681741714 | 461   | +3%        |  0.0000 |
+| TPC-H 08       | 4.90201091766  | 295   | 5.89004468918 | 354   | +20%       |  0.0000 |
 | TPC-H 09       | 1.49427366257  | 90    | 1.48768651485 | 90    | -0%        |  0.2871 |
 | TPC-H 10       | 2.8550825119   | 172   | 2.89195108414 | 174   | +1%        |  0.1555 |
 | TPC-H 11       | 28.5678386688  | 1715  | 28.8567256927 | 1732  | +1%        |  0.0000 |
 | TPC-H 12       | 8.83416938782  | 531   | 9.23126125336 | 554   | +4%        |  0.0000 |
 | TPC-H 13       | 2.20315814018  | 133   | 2.23280572891 | 134   | +1%        |  0.0000 |
 | TPC-H 14       | 46.2855300903  | 2778  | 46.6186714172 | 2798  | +1%        |  0.0000 |
 | TPC-H 15       | 72.3205032349  | 4340  | 71.6943283081 | 4302  | -1%        |  0.0000 |
 | TPC-H 16       | 7.01711797714  | 422   | 7.11375999451 | 427   | +1%        |  0.0000 |
 | TPC-H 17       | 1.53186941147  | 92    | 1.56167197227 | 94    | +2%        |  0.0000 |
 | TPC-H 18       | 0.796154022217 | 48    | 0.77923309803 | 47    | -2%        |  0.0243 |
+| TPC-H 19       | 32.2467460632  | 1935  | 34.5593261719 | 2074  | +7%        |  0.0000 |
 | TPC-H 20       | 2.73897361755  | 165   | 2.64484596252 | 159   | -3%        |  0.0000 |
 | TPC-H 21       | 1.42658638954  | 86    | 1.46430814266 | 88    | +3%        |  0.0000 |
 | TPC-H 22       | 14.2510633469  | 856   | 13.883898735  | 834   | -3%        |  0.0000 |
 | geometric mean |                |       |               |       | +1%        |         |
 +----------------+----------------+-------+---------------+-------+------------+---------+
```

This is what it looks like for a single cpp:
```diff
 +----------------+----------------+-------+----------------+-------+------------+---------+
 | Benchmark      | prev. iter/s   | runs  | new iter/s     | runs  | change [%] | p-value |
 +----------------+----------------+-------+----------------+-------+------------+---------+
+| TPC-H 01       | 0.599761068821 | 36    | 0.633876204491 | 39    | +6%        |  0.0000 |
+| TPC-H 02       | 11.8571109772  | 712   | 12.5952978134  | 756   | +6%        |  0.0000 |
 | TPC-H 03       | 6.96436452866  | 418   | 7.21092510223  | 433   | +4%        |  0.0000 |
 | TPC-H 04       | 6.85212564468  | 412   | 6.95795297623  | 418   | +2%        |  0.0010 |
 | TPC-H 05       | 4.43019294739  | 266   | 4.29929304123  | 258   | -3%        |  0.0000 |
 | TPC-H 06       | 195.011672974  | 11701 | 198.41897583   | 11906 | +2%        |  0.0000 |
 | TPC-H 07       | 7.42980623245  | 446   | 7.62684011459  | 458   | +3%        |  0.0000 |
+| TPC-H 08       | 4.90201091766  | 295   | 5.32516765594  | 320   | +9%        |  0.0000 |
 | TPC-H 09       | 1.49427366257  | 90    | 1.52247714996  | 92    | +2%        |  0.0001 |
 | TPC-H 10       | 2.8550825119   | 172   | 2.93201828003  | 176   | +3%        |  0.0020 |
 | TPC-H 11       | 28.5678386688  | 1715  | 29.3544807434  | 1762  | +3%        |  0.0000 |
 | TPC-H 12       | 8.83416938782  | 531   | 8.64130973816  | 519   | -2%        |  0.0000 |
+| TPC-H 13       | 2.20315814018  | 133   | 2.34939694405  | 141   | +7%        |  0.0000 |
 | TPC-H 14       | 46.2855300903  | 2778  | 47.0677833557  | 2825  | +2%        |  0.0000 |
 | TPC-H 15       | 72.3205032349  | 4340  | 72.108215332   | 4327  | -0%        |  0.0003 |
 | TPC-H 16       | 7.01711797714  | 422   | 7.32353448868  | 440   | +4%        |  0.0000 |
 | TPC-H 17       | 1.53186941147  | 92    | 1.54534137249  | 93    | +1%        |  0.0103 |
+| TPC-H 18       | 0.796154022217 | 48    | 0.848833739758 | 51    | +7%        |  0.0000 |
 | TPC-H 19       | 32.2467460632  | 1935  | 32.4772758484  | 1949  | +1%        |  0.0000 |
 | TPC-H 20       | 2.73897361755  | 165   | 2.84375214577  | 171   | +4%        |  0.0000 |
 | TPC-H 21       | 1.42658638954  | 86    | 1.43206572533  | 86    | +0%        |  0.2658 |
 | TPC-H 22       | 14.2510633469  | 856   | 14.1954784393  | 852   | -0%        |  0.2322 |
 | geometric mean |                |       |                |       | +3%        |         |
 +----------------+----------------+-------+----------------+-------+------------+---------+
```

Thus, we should make sure that we do not mix unity and non-unity builds.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -45,4 +45,10 @@ RUN apt-get update \
     && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* \
     && ln -sf /usr/bin/llvm-symbolizer-3.8 /usr/bin/llvm-symbolizer
 
+# Temporarily for cmake release candidate:
```
Comment (tobodner) [inline Dockerfile]: Consider adding a TODO.

**Comment by mrks** [comment]
Closing this for now, as we wait for cmake 3.16 to land

File: src/test/sql/sql_pipeline_test.cpp [hunk 2]
```diff
@@ -18,21 +18,6 @@
 #include "sql/sql_pipeline_builder.hpp"
 #include "sql/sql_plan_cache.hpp"
 
-namespace {
```
Comment (mrks) [inline src/test/sql/sql_pipeline_test.cpp]: unused

File: README.md [hunk 3]
```diff
@@ -57,8 +57,11 @@ Subsequent calls to CMake, e.g., when adding files to the build will not be nece
...
+
 ### ccache
-For development, we strongly suggest to use [ccache](https://ccache.samba.org/), which reduces the time needed for recompiles significantly. Especially when switching branches, this can reduce the time to recompile from several minutes to one or less. To use ccache, simply add `-DCMAKE_CXX_COMPILER_LAUNCHER=ccache` to your cmake call.
+For development, we suggest to use [ccache](https://ccache.samba.org/), which reduces the time needed for recompiles significantly. Especially when switching branches, this can reduce the time to recompile from several minutes to one or less. To use ccache, add `-DCMAKE_CXX_COMPILER_LAUNCHER=ccache` to your cmake call. You will need to [adjust some ccache settings](https://ccache.dev/manual/latest.html#_precompiled_headers) to that it works with precompiled headers. On our CI server, this works for us: `CCACHE_SLOPPINESS=file_macro,pch_defines,time_macros CCACHE_DEPEND=1`.
```
Comment (Bouncner) [inline README.md]: Is this going to change or do I need to add this sloppiness each time a run cmake now?

File: src/test/statistics/chunk_statistics/range_filter_test.cpp [hunk 4]
```diff
@@ -1,278 +0,0 @@
...
-#include "statistics/chunk_statistics/range_filter.hpp"
-#include "types.hpp"
-
-namespace opossum {
-
-template <typename T>
-class RangeFilterTest : public ::testing::Test {
- protected:
-  void SetUp() override {
-    // Manually created vector. Largest exlusive gap (only gap when gap_count == 1) will
```
Comment (Bouncner) [inline src/test/statistics/chunk_statistics/range_filter_test.cpp]: Why removed?

File: src/lib/optimizer/strategy/column_pruning_rule.cpp [hunk 5]
```diff
@@ -27,28 +27,28 @@ namespace {
 void gather_expressions_not_computed_by_expression_evaluator(
     const std::shared_ptr<AbstractExpression>& expression,
     const std::vector<std::shared_ptr<AbstractExpression>>& input_expressions,
-    ExpressionUnorderedSet& required_expressions, const bool top_level = true) {
+    ExpressionUnorderedSet& REKWAIRED_EXPRESSION, const bool top_level = true) {
   // Top-level expressions are those that are (part of) the ExpressionEvaluator's final result. For example, for an
   // ExpressionEvaluator producing (a + b) + c, the entire expression is a top-level expression. It is the consumer's
   // job to mark it as required. (a + b) however, is required by the ExpressionEvaluator and will be added to
-  // required_expressions, as it is not a top-level expression.
+  // REKWAIRED_EXPRESSION, as it is not a top-level expression.
```
Comment (Bouncner) [inline src/lib/optimizer/strategy/column_pruning_rule.cpp]: ?

Comment (mrks) [reply src/lib/optimizer/strategy/column_pruning_rule.cpp]: Wanted to make sure that clang-tidy still shouts at me even when we use unity build (see https://gitlab.kitware.com/cmake/cmake/issues/20058)

Comment (mrks) [reply src/lib/optimizer/strategy/column_pruning_rule.cpp]: Will be removed once this is ready to go (and Jenkins will make sure that I do)

File: src/test/statistics/chunk_statistics/range_filter_test.cpp (hunk 4 above)
Comment (mrks) [reply src/test/statistics/chunk_statistics/range_filter_test.cpp]: There is another version in `statistics_objects/range_filter_test.cpp`. CMake doesn't even know about this one.

File: README.md (hunk 3 above)
Comment (mrks) [reply README.md]: There are different ways to set ccache options, for example in your environment variables or in . I'll add a sentence pointing to https://ccache.dev/manual/latest.html#_configuration

**Comment by mrks** [comment]
This should be ready to go once cmake 3.16 lands in homebrew. Could you have a look please?

**Comment by mrks** [comment]
Finally green :)

**Comment by Bensk1** [comment]
Any news here?

**Comment by mrks** [comment]
When setting up my new Macbook, I decided to test if Xcode is really necessary. It is not - homebrew already makes sure the command line tools are installed and that is all we need.

We should buy new hardware more often to test things like this!

**Comment by Bensk1** [comment]
> We should buy new hardware more often to test things like this!

Agreed!

**Comment by Bouncner** [comment]
Is anybody on this? I am not familiar with CMake so there might be better reviewers. But it would be nice to get this PR finally merged.

**Comment by tobodner** [comment]
> Is anybody on this? I am not familiar with CMake so there might be better reviewers. But it would be nice to get this PR finally merged.

Yes, this in on my list for today.

**Comment by tobodner** [comment]
> * Uses gold instead of ld under Linux

Under Linux, you are now using LLVM LLD instead of GNU Gold.

**Comment by mrks** [comment]
fixed

File: src/test/CMakeLists.txt [hunk 6]
```diff
@@ -244,6 +244,25 @@ add_dependencies(hyriseTest hyriseTestPlugin hyriseTestNonInstantiablePlugin)
...
+        <gtest/gtest.h>
+        [["all_parameter_variant.hpp"]]
+        [["storage/create_iterable_from_segment.hpp"]]
+        [["storage/table.hpp"]]
+        [["types.hpp"]]
+    )
+
+    # Exclude files that use gmock as these are incompatible with unity builds and pre-compiled headers
+    set_source_files_properties(
+        server/server_session_test.cpp
```
Comment (tobodner) [inline src/test/CMakeLists.txt]: Please merge a recent master into this branch. This file is not part of the server implementation anymore. Also, AFAIK, we got more mock classes in our code base.

File: CMakeLists.txt [hunk 7]
```diff
@@ -92,17 +107,17 @@ set(CMAKE_SHARED_LINKER_FLAGS_DEBUG "${CMAKE_SHARED_LINKER_FLAGS_DEBUG} -O0")
 set(CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG} -O0")
 
 # Dependencies
-find_package(Numa)
+find_package(Numa QUIET)
```
[the remark this answers is not in the document]
Comment (mrks) [reply CMakeLists.txt]: Stops find_package from complaining if it does not find the package

File: Jenkinsfile [hunk 8]
```diff
@@ -48,40 +48,54 @@ try {
     def oppossumCI = docker.image('hyrise/opossum-ci:19.10');
     oppossumCI.pull()
     // create ccache volume on host using:
-    // mkdir /mnt/ccache; mount -t tmpfs -o size=50G none /mnt/ccache
+    // mkdir /mnt/ccache; mount -t tmpfs -o size=200G none /mnt/ccache
```
[the remark this answers is not in the document]
Comment (mrks) [reply Jenkinsfile]: Nah, as random as the one before, just bigger.

File: CMakeLists.txt [hunk 9]
```diff
@@ -82,6 +84,19 @@ set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} ${FLAGS_ALL} ${FLAGS_REL
...
+    if(DEFINED CMAKE_UNITY_BUILD)
+        message(FATAL_ERROR "-DCMAKE_UNITY_BUILD=ON was set, but your cmake version does not support unity builds.")
+    endif()
+endif()
+
+# On Linux, use the lld linker, which is twice as fast for incremental debug builds
+if (NOT APPLE)
+    set(CMAKE_SHARED_LINKER_FLAGS "-fuse-ld=lld")
+    set(CMAKE_EXE_LINKER_FLAGS "-fuse-ld=lld")
+endif()
```
[the remark this answers is not in the document]
Comment (mrks) [reply CMakeLists.txt]: done

**Comment by tobodner** [comment]
Would you please ping me, when I can make a final check?

**Comment by mrks** [comment]
Sure. So far it is just back and forth, trying to understand what exactly causes the CI fails. It doesn't help that I am unable to reproduce this on my own.

**Comment by mrks** [comment]
Doesn't seem to be caused by #1966. Not sure if that is a good thing, because now I am left without ideas again.

**Comment by mrks** [comment]
Ok, I managed to break things on nemea without this PR. I reset the branch to be just like it was when it was approved.

**Pull request merged**